### PR TITLE
fix transformers version check for InternVL2

### PR DESCRIPTION
**Pull request by zhyncs**
## Motivation

hot fix cc @AllentDan @RunningLeon @grimoire @lvhan028 

ref https://huggingface.co/OpenGVLab/InternVL2-26B/blob/main/config.json#L14

## Modification

as titled

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
3. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by zhyncs** [comment]
In pr https://github.com/InternLM/lmdeploy/pull/1917, I verified [Qwen 2 7B Instruct](https://huggingface.co/Qwen/Qwen2-7B-Instruct/blob/main/config.json#L23) and [DeepSeek VL 1.3b Chat](https://huggingface.co/deepseek-ai/deepseek-vl-1.3b-chat/blob/main/config.json#L28) at that time. For these two models, whether it's LLM or VLM, the `transformers_version` field is located in the config.json **top-level** directory. However, in [InternVL2](https://huggingface.co/OpenGVLab/InternVL2-26B/blob/main/config.json#L14), it is located in the `llm_config` **secondary** directory. I don't know if this kind of fix can cover all situations currently. Do you have any suggestions? @grimoire

File: lmdeploy/tokenizer.py [hunk 1]
```diff
@@ -231,6 +231,8 @@ def _check_transformers_version(self, model_dir: str):
 
         current_transformers_version = version.parse(transformers.__version__)
         cfg = get_model_arch(model_dir)[1]
+        if getattr(cfg, 'transformers_version', 'str') is None:
+            cfg = getattr(cfg, 'llm_config')
         required_transformers_version = version.parse(
```
Comment (zhyncs) [inline lmdeploy/tokenizer.py:236]: Do I need to change this line to these?
```python3
        cfg_ver = getattr(cfg, 'transformers_version', 'str')
        if cfg_ver is None:
            return
        required_transformers_version = version.parse(cfg_ver)
```

Comment (RunningLeon) [reply lmdeploy/tokenizer.py:236]: yes, should be `None` instead of `'str'`

Comment (zhyncs) [reply lmdeploy/tokenizer.py:236]: ok

**Comment by grimoire** [comment]
> Do you have any suggestions?

It is OK to add a check like this PR.
I might ignore the check for certain models that can not read the version with `AutoConfig.from_pretrained`. Adding too many special cases is annoying.

**Review by grimoire — APPROVED** [review]

**Comment by zhyncs** [comment]
> > Do you have any suggestions?
> 
> It is OK to add a check like this PR. I might ignore the check for certain models that can not read the version with `AutoConfig.from_pretrained`. Adding too many special cases is annoying.

ok

File: lmdeploy/tokenizer.py (hunk 1 above)
Comment (zhyncs) [reply lmdeploy/tokenizer.py]: done cc @RunningLeon

**Comment by zhyncs** [comment]
cc @RunningLeon @lvhan028

File: lmdeploy/tokenizer.py [hunk 2]
```diff
@@ -231,8 +231,14 @@ def _check_transformers_version(self, model_dir: str):
 
         current_transformers_version = version.parse(transformers.__version__)
         cfg = get_model_arch(model_dir)[1]
-        required_transformers_version = version.parse(
-            getattr(cfg, 'transformers_version', 'str'))
+        cfg_ver = getattr(cfg, 'transformers_version', None)
```
Comment (lvhan028) [inline lmdeploy/tokenizer.py:234]: Hi, @zhyncs 
Would you mind adding unitests for this function?

Comment (zhyncs) [reply lmdeploy/tokenizer.py:234]: ok

Comment (zhyncs) [reply lmdeploy/tokenizer.py:234]: The assertion ensures that the tokenizer is not `None`, confirming its successful creation and passing the check.

**Comment by zhyncs** [comment]
I'll take a look. Since the development machine does not have external network access, I will see how to solve this.

**Comment by zhyncs** [comment]
transformers version: 4.40.0

```python3
import pytest

from lmdeploy.tokenizer import HuggingFaceTokenizer

@pytest.mark.parametrize('model_path', [
    'Qwen/Qwen2-0.5B-Instruct'
])
def test_check_transformers_version(model_path):
    tokenizer = HuggingFaceTokenizer(model_path)
    assert tokenizer is not None
```

```
=========================================================== test session starts ============================================================
platform linux -- Python 3.9.16, pytest-8.2.2, pluggy-1.5.0
rootdir: /workdir/lmdeploy
plugins: hydra-core-1.2.0, anyio-3.7.1
collected 1 item

tests/test_lmdeploy/test_tokenizer.py .                                                                                              [100%]

============================================================= warnings summary =============================================================
../../usr/local/lib/python3.9/site-packages/yapf/yapflib/errors.py:16
  /usr/local/lib/python3.9/site-packages/yapf/yapflib/errors.py:16: PendingDeprecationWarning: lib2to3 package is deprecated and may not be able to parse Python 3.10+
    from lib2to3.pgen2 import tokenize

tests/test_lmdeploy/test_tokenizer.py::test_check_transformers_version[Qwen/Qwen2-0.5B-Instruct]
  /usr/local/lib/python3.9/site-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
    warnings.warn(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
====================================================== 1 passed, 2 warnings in 34.23s ======================================================
```

**Comment by zhyncs** [comment]
https://huggingface.co/Qwen/Qwen2-0.5B-Instruct/blob/main/config.json#L23

In the Qwen 2 0.5B Instruct, there is a field called `transformers_version`, and the logic of llm_config get will not be executed.

**Comment by zhyncs** [comment]
double check locally
```
[root@hostname lmdeploy]# cat local_test.py
from lmdeploy.tokenizer import HuggingFaceTokenizer
tokenizer = HuggingFaceTokenizer('/workdir/Qwen2-0.5B-Instruct')
assert tokenizer is not None

[root@hostname lmdeploy]# python3 local_test.py
4.40.1
2024-07-08 22:13:45,095 - lmdeploy - WARNING - The current version of `transformers` is transformers==4.40.0, which is lower than the required version transformers==4.40.1. Please upgrade to the required version.
Special tokens have been added in the vocabulary, make sure the associated word embeddings are fine-tuned or trained.

[root@hostname lmdeploy]# echo $?
0
```

**Comment by zhyncs** [comment]
I know the reason now. The `llm_config` for [InternVL2 1B](https://huggingface.co/OpenGVLab/InternVL2-1B/blob/main/config.json#L14) is
`<class'transformers.models.qwen2.configuration_qwen2.Qwen2Config'>`, and for [InternVL2 26B](https://huggingface.co/OpenGVLab/InternVL2-26B/blob/main/config.json#L14) it's a `dict`.
It's interesting lol. @lvhan028 @grimoire @zhulinJulia24 
And why [InternVL2 1B](https://huggingface.co/OpenGVLab/InternVL2-1B/blob/main/config.json#L14) use the `Qwen/Qwen2-0.5B-Instruct` as the `_name_or_path`.

**Comment by zhyncs** [comment]
> Adding too many special cases is annoying.

@grimoire You are right. This type of configuration is indeed a bit troublesome. I took the easy way out and directly copied the `config.json` from InternVL2-26B to deepseek-vl-1.3b-chat for verification, and it passed locally but it's wrong. In order to verify its **real correctness**, it is still necessary to download the model to the local environment and then I find the root cause.

https://github.com/InternLM/lmdeploy/actions/runs/9842035941/job/27170079420?pr=1952 Unit Test finally passed😂 @lvhan028 The lint failure is due to a 503 error on a URL, please ignore for now.

**Review by AllentDan — APPROVED** [review]

**Pull request merged**